File: palworld_save_pal/game/uid_swap.py
from typing import Any, Callable, Dict, List, Optional
from uuid import UUID

from palworld_save_pal.game.pal_objects import PalObjects
from palworld_save_pal.utils.logging_config import create_logger
from palworld_save_pal.utils.uuid import are_equal_uuids

logger = create_logger(__name__)

OWNERSHIP_KEYS = (
    "OwnerPlayerUId",
    "owner_player_uid",
    "build_player_uid",
    "private_lock_player_uid",
)


def _deep_swap_uids(data: Any, old_uid: str, new_uid: str) -> None:
    if isinstance(data, dict):
        for key in OWNERSHIP_KEYS:
            val = data.get(key)
            if val is None:
                continue
            if isinstance(val, dict):
                inner = val.get("value")
                if isinstance(inner, str) and inner.lower() == old_uid:
                    val["value"] = new_uid
                elif isinstance(inner, str) and inner.lower() == new_uid:
                    val["value"] = old_uid
            elif isinstance(val, str):
                if val.lower() == old_uid:
                    data[key] = new_uid
                elif val.lower() == new_uid:
                    data[key] = old_uid
        for v in data.values():
            _deep_swap_uids(v, old_uid, new_uid)
    elif isinstance(data, list):
        for item in data:
            _deep_swap_uids(item, old_uid, new_uid)


async def swap_player_uids(
    save_manager,
    old_player_uid: UUID,
    new_player_uid: UUID,
    ws_callback: Optional[Callable] = None,
) -> Dict[str, Any]:
    async def progress(msg: str):
        logger.info(msg)
        if ws_callback:
            await ws_callback(msg)

    old_uid_str = str(old_player_uid).lower()
    new_uid_str = str(new_player_uid).lower()

    if old_uid_str == new_uid_str:
        return {"error": "Both players are the same."}

    await progress("Validating players...")

    # Validate both players have file refs (i.e. exist as players with .sav files)
    if old_player_uid not in save_manager._player_file_refs:
        return {"error": f"Player {old_player_uid} not found in save."}
    if new_player_uid not in save_manager._player_file_refs:
        return {"error": f"Player {new_player_uid} not found in save."}

    # Load player GVAS files if not already loaded
    if old_player_uid not in save_manager._player_gvas_files:
        await save_manager.load_player_on_demand(old_player_uid)
    if new_player_uid not in save_manager._player_gvas_files:
        await save_manager.load_player_on_demand(new_player_uid)

    old_gvas = save_manager._player_gvas_files.get(old_player_uid)
    new_gvas = save_manager._player_gvas_files.get(new_player_uid)
    if not old_gvas or not new_gvas:
        return {"error": "Failed to load player save files."}

    # Validate level >= 2 using summaries (level is in CharacterSaveParameterMap, not player GVAS)
    old_summary = save_manager._player_summaries.get(old_player_uid)
    new_summary = save_manager._player_summaries.get(new_player_uid)
    old_level = (old_summary.level or 1) if old_summary else 1
    new_level = (new_summary.level or 1) if new_summary else 1
    if old_level < 2 or new_level < 2:
        return {
            "error": f"Both players must be at least level 2. "
            f"Player 1 is level {old_level}, Player 2 is level {new_level}."
        }

    # Get save data from player GVAS files
    old_save_data = PalObjects.get_value(old_gvas.sav.properties["SaveData"])
    new_save_data = PalObjects.get_value(new_gvas.sav.properties["SaveData"])

    # Get instance IDs
    old_instance_id = PalObjects.get_guid(
        old_save_data["IndividualId"]["value"]["InstanceId"]
    )
    new_instance_id = PalObjects.get_guid(
        new_save_data["IndividualId"]["value"]["InstanceId"]
    )

    await progress("Swapping player UIDs in save data...")

    # Step 1: Swap PlayerUId in player GVAS files
    PalObjects.set_value(old_save_data["PlayerUId"], new_uid_str)
    PalObjects.set_value(
        old_save_data["IndividualId"]["value"]["PlayerUId"], new_uid_str
    )
    PalObjects.set_value(new_save_data["PlayerUId"], old_uid_str)
    PalObjects.set_value(
        new_save_data["IndividualId"]["value"]["PlayerUId"], old_uid_str
    )

    await progress("Swapping UIDs in character save parameter map...")

    # Step 2: Swap PlayerUId in CharacterSaveParameterMap
    for entry in save_manager._character_save_parameter_map:
        try:
            instance_id = PalObjects.get_guid(entry["key"]["InstanceId"])
            if are_equal_uuids(instance_id, old_instance_id):
                PalObjects.set_value(entry["key"]["PlayerUId"], new_uid_str)
            elif are_equal_uuids(instance_id, new_instance_id):
                PalObjects.set_value(entry["key"]["PlayerUId"], old_uid_str)
        except (KeyError, TypeError):
            continue

    await progress("Swapping UIDs in guild data...")

    # Step 3: Swap in guild data
    for group_entry in save_manager._group_save_data_map:
        try:
            group_type_val = PalObjects.get_enum_property(
                PalObjects.get_nested(group_entry, "value", "GroupType")
            )
            if "Guild" not in str(group_type_val):
                continue
        except (KeyError, TypeError):
            continue

        try:
            raw_data = group_entry["value"]["RawData"]["value"]
        except (KeyError, TypeError):
            continue

        # Swap in individual_character_handle_ids
        for handle in raw_data.get("individual_character_handle_ids", []):
            handle_guid = handle.get("guid")
            handle_inst = handle.get("instance_id")
            if handle_inst and are_equal_uuids(handle_inst, old_instance_id):
                handle["guid"] = new_uid_str
            elif handle_inst and are_equal_uuids(handle_inst, new_instance_id):
                handle["guid"] = old_uid_str

        # Swap admin_player_uid
        admin_uid = raw_data.get("admin_player_uid")
        if admin_uid:
            admin_str = str(admin_uid).lower() if not isinstance(admin_uid, str) else admin_uid.lower()
            if admin_str == old_uid_str:
                raw_data["admin_player_uid"] = new_uid_str
            elif admin_str == new_uid_str:
                raw_data["admin_player_uid"] = old_uid_str

        # Swap in players list
        for player_entry in raw_data.get("players", []):
            p_uid = player_entry.get("player_uid")
            if p_uid:
                p_str = str(p_uid).lower() if not isinstance(p_uid, str) else p_uid.lower()
                if p_str == old_uid_str:
                    player_entry["player_uid"] = new_uid_str
                elif p_str == new_uid_str:
                    player_entry["player_uid"] = old_uid_str

    await progress("Swapping ownership references across all data...")

    # Step 4: Deep recursive swap across all level data
    _deep_swap_uids(save_manager._gvas_file.properties, old_uid_str, new_uid_str)

    await progress("Swapping player file references...")

    # Step 5: Swap DPS data if present
    old_dps = old_gvas.dps
    new_dps = new_gvas.dps
    # Swap the DPS references between the two PlayerGvasFiles
    old_gvas.dps = new_dps
    new_gvas.dps = old_dps

    # Step 6: Swap player GVAS files and file refs in SaveManager dicts
    save_manager._player_gvas_files[old_player_uid] = new_gvas
    save_manager._player_gvas_files[new_player_uid] = old_gvas

    old_file_ref = save_manager._player_file_refs.get(old_player_uid)
    new_file_ref = save_manager._player_file_refs.get(new_player_uid)
    save_manager._player_file_refs[old_player_uid] = new_file_ref
    save_manager._player_file_refs[new_player_uid] = old_file_ref

    await progress("Rebuilding caches...")

    # Step 7: Invalidate all caches and rebuild summaries
    save_manager.invalidate_performance_caches()
    save_manager._players.clear()
    save_manager._guilds.clear()
    save_manager._loaded_players.clear()
    save_manager._loaded_guilds.clear()
    save_manager._extract_player_summaries()
    save_manager._extract_guild_summaries()

    await progress("UID swap complete!")

    return {"success": True}

File: rust/psp-core/src/domain/uid_swap.rs
//! Cross-player UID swap -- port of `game/mixins/player_swap.py`'s
//! `PlayerSwapMixin.swap_player_uids` and its private helpers
//! (`_validate_swap_players`, `_swap_player_gvas_uids`,
//! `_swap_character_save_parameters`, `_swap_guild_member_uids` +
//! `_swap_guild_character_handles`/`_swap_guild_admin_uid`/
//! `_swap_guild_player_list`, `_swap_player_file_refs`).
//!
//! Reconciliation with the brief: the brief's assumed interface
//! (`CharacterMapEntry::set_key_player_uid`, `GroupMapEntry::
//! swap_player_uids`, `SaveSession::player_file_refs()`/`player_gvas()`/
//! `character_map_mut()`/`level_properties_mut()` as zero-arg METHODS) does
//! not exist as such -- there are no typed wrapper structs over
//! `uesave::MapEntry` in this port (Task 3E-2/transfer.rs already
//! established this). This module is written directly against the same
//! real navigation layer transfer.rs uses: `domain::world`'s
//! character/group-map accessors and entry helpers (now including this
//! task's own `world::set_entry_player_uid` write helper),
//! `domain::player`'s `save_data_props`, `domain::guild_tail`'s
//! `GuildTail`/`entry_group_type`/`entry_group_data_mut` raw-tail codec,
//! and `transfer::ensure_player_gvas_loaded`/`transfer::
//! save_data_instance_id` (both promoted to `pub(crate)` by this task so
//! this module can reuse them verbatim instead of re-implementing on-demand
//! GVAS loading a second time). `SaveSession::player_file_refs` and
//! `SaveSession::loaded_players` are plain fields, not methods -- called as
//! such throughout. The `level_properties_mut`/`swap_player_file_refs`/
//! `swap_player_gvas_uids` session methods ARE real (this task adds them to
//! `session.rs`, matching the brief's exact signatures).
//!
//! The brief's own literal `SaveSession::swap_player_uids` code block (its
//! Step 3) omits `_validate_swap_players`'s level>=2 gate and its "Failed to
//! load player save files."/"Player {uid} not found in save." wording --
//! reproduced here exactly as the brief's code gives it (bare "Player {uid}
//! not found.", no level check), not as `player_swap.py` gives it. The
//! brief's progress-string list is explicitly scoped to `player_swap.py:
//! 79-120`, which also excludes line 123's trailing `"UID swap complete!"`
//! progress message -- also not reproduced here, matching that scoping.

use crate::domain::guild_tail;
use crate::domain::{player, world};
use crate::error::CoreError;
use crate::progress::ProgressSink;
use crate::props;
use crate::session::SaveSession;
use crate::transfer::{self, TransferError};
use uuid::Uuid;

/// `_OWNERSHIP_KEYS` (`player_swap.py:19-24`).
const OWNERSHIP_KEYS: [&str; 4] = [
    "OwnerPlayerUId",
    "owner_player_uid",
    "build_player_uid",
    "private_lock_player_uid",
];

/// Port of `_swap_guild_member_uids` (`player_swap.py:191-263`), folding its
/// three sub-helpers (`_swap_guild_character_handles`/`_swap_guild_admin_uid`/
/// `_swap_guild_player_list`) into one pass per Guild-type
/// `GroupSaveDataMap` entry, exactly as Python calls all three per entry:
///
/// * `individual_character_handle_ids` (a typed `PalGroupData` field, not
///   part of the raw guild tail -- see `domain::guild_tail`'s own doc
///   comment on the RawData split) -- any handle whose `instance_id`
///   matches either player's own character instance id has its `guid`
///   field retargeted to the OTHER player;
/// * the tail's `admin_player_uid`;
/// * every tail `players[].player_uid`.
///
/// Each swapped bidirectionally (old->new, new->old), matching Python's
/// `are_equal_uuids(x, old) -> new` / `are_equal_uuids(x, new) -> old`
/// pattern. Python's own guard is `"Guild" not in str(group_type): continue`
/// (a substring check); this instead matches the exact enum variant name
/// `"EPalGroupType::Guild"`, the same precision `transfer.rs`'s own guild
/// code already established for this exact check.
fn swap_guild_member_uids(
    level: &mut uesave::Save,
    old_uid: Uuid,
    new_uid: Uuid,
    old_instance_id: Uuid,
    new_instance_id: Uuid,
) -> Result<(), CoreError> {
    for entry in world::group_map_mut(level)?.iter_mut() {
        if guild_tail::entry_group_type(entry).as_deref() != Some("EPalGroupType::Guild") {
            continue;
        }
        let Some(group_data) = guild_tail::entry_group_data_mut(entry) else {
            continue;
        };

        for handle in group_data.individual_character_handle_ids.iter_mut() {
            let handle_instance_id = props::guid_to_uuid(&handle.instance_id);
            if handle_instance_id == old_instance_id {
                handle.guid = props::uuid_to_guid(new_uid);
            } else if handle_instance_id == new_instance_id {
                handle.guid = props::uuid_to_guid(old_uid);
            }
        }

        // The admin uid and every member `player_uid` are swapped
        // bidirectionally in place; uesave re-serializes the structured guild
        // on save, and all other tail fields (roles, names, markers) survive
        // untouched.
        let Some(guild) = guild_tail::as_guild_mut(group_data) else {
            continue;
        };
        guild_tail::swap_player_uids(guild, old_uid, new_uid);
    }
    Ok(())
}

impl SaveSession {
    /// Port of `PlayerSwapMixin.swap_player_uids` (`player_swap.py:61-124`).
    /// `Ok(())` == Python's `{"success": True}`;
    /// `Err(TransferError::Rejected(msg))` == Python's `{"error": msg}` (a
    /// SOFT rejection -- the handler maps this to a normal `{"error": ...}`
    /// response on the `swap_player_uids` wire type, never the hard WS
    /// `error` frame). `TransferError` is reused verbatim from Task 3E-2's
    /// `transfer.rs`, which already carries this exact soft/hard split.
    pub fn swap_player_uids(
        &mut self,
        old_player_uid: Uuid,
        new_player_uid: Uuid,
        progress: &ProgressSink,
    ) -> Result<(), TransferError> {
        if old_player_uid == new_player_uid {
            return Err(TransferError::Rejected("Both players are the same.".into()));
        }

        progress("Validating players...");
        for uid in [old_player_uid, new_player_uid] {
            if !self.player_file_refs.contains_key(&uid) {
                return Err(TransferError::Rejected(format!("Player {uid} not found.")));
            }
            transfer::ensure_player_gvas_loaded(self, uid)?;
        }

        let old_instance_id = self
            .loaded_players
            .get(&old_player_uid)
            .and_then(|loaded| player::save_data_props(&loaded.sav).ok())
            .and_then(transfer::save_data_instance_id)
            .ok_or_else(|| {
                TransferError::Rejected("Source player SaveData is missing or invalid.".into())
            })?;
        let new_instance_id = self
            .loaded_players
            .get(&new_player_uid)
            .and_then(|loaded| player::save_data_props(&loaded.sav).ok())
            .and_then(transfer::save_data_instance_id)
            .ok_or_else(|| {
                TransferError::Rejected("Target player SaveData is missing or invalid.".into())
            })?;

        progress("Swapping player UIDs in save data...");
        self.swap_player_gvas_uids(old_player_uid, new_player_uid);

        progress("Swapping UIDs in character save parameter map...");
        for entry in world::character_map_mut(&mut self.level)?.iter_mut() {
            let Some(instance_id) = world::entry_instance_id(entry) else {
                continue;
            };
            if instance_id == old_instance_id {
                world::set_entry_player_uid(entry, new_player_uid);
            } else if instance_id == new_instance_id {
                world::set_entry_player_uid(entry, old_player_uid);
            }
        }

        progress("Swapping UIDs in guild data...");
        swap_guild_member_uids(
            &mut self.level,
            old_player_uid,
            new_player_uid,
            old_instance_id,
            new_instance_id,
        )?;

        progress("Swapping ownership references across all data...");
        props::swap_uuid_values_deep(
            self.level_properties_mut(),
            &OWNERSHIP_KEYS,
            old_player_uid,
            new_player_uid,
        );

        progress("Swapping player file references...");
        self.swap_player_file_refs(old_player_uid, new_player_uid);

        progress("Rebuilding caches...");
        self.rebuild_player_caches()?;

        Ok(())
    }
}
